playground/engine.py

```diff
@@ -0,0 +1,818 @@
+                    device=device
+                )
+            case _:
+                mesh_path = self.assets[geometry_type]
+                mesh = load_mesh(mesh_path, device)
+                materials = load_materials(mesh, device)
+                if len(materials) > 0:
+                    load_missing_material_info(mesh_path, materials, device)
+                    material_index_mapping = self.register_materials(materials=materials, model_name=geometry_type)
+                    # Update material assignments to match playground material registry
+                    material_index_mapping = material_index_mapping.to(device=device)
+                    material_id = mesh.material_assignments.to(device=device, dtype=torch.long)
+                    mesh.material_assignments = material_index_mapping[material_id].int()
+        # Always use default material, if no materials were specified
+        mesh.material_assignments = torch.max(mesh.material_assignments, torch.zeros_like(mesh.material_assignments))
+        return mesh
+
+    def recompute_stacked_buffers(self):
+        objects = [p.apply_transform() for p in self.objects.values()]
+        # Recompute primitive type tensor
+        for obj in objects:
+            f = obj.triangles
+            num_faces = f.shape[0]
+            obj.primitive_type_tensor = f.new_full(size=(num_faces,), fill_value=obj.primitive_type.value)
+            obj.refractive_index_tensor = f.new_full(size=(num_faces,),
+                                                     fill_value=obj.refractive_index, dtype=torch.float)
+
+        # Stack fields again
+        self.stacked_fields = None
+        if self.has_visible_objects():
+            self.stacked_fields = OptixPrimitive.stack([p for p in objects if p.primitive_type != OptixPrimitiveTypes.NONE])
+
+    def has_visible_objects(self):
+        return len([p for p in self.objects.values() if p.primitive_type != OptixPrimitiveTypes.NONE]) > 0
+
+    @torch.cuda.nvtx.range("rebuild_bvh (prim)")
+    def rebuild_bvh_if_needed(self, force=False, rebuild=True):
+        if self.dirty or force:
+            if self.has_visible_objects():
+                self.recompute_stacked_buffers()
+                self.tracer.build_mesh_acc(
+                    mesh_vertices=self.stacked_fields.vertices,
+                    mesh_faces=self.stacked_fields.triangles,
+                    rebuild=rebuild,
+                    allow_update=True
+                )
+            else:
+                self.tracer.build_mesh_acc(
+                    mesh_vertices=torch.zeros([3, 3], dtype=torch.float, device='cuda'),
+                    mesh_faces=torch.zeros([1, 3], dtype=torch.int, device='cuda'),
+                    rebuild=True,
+                    allow_update=True
+                )
+        self.dirty = False
+
+#################################
+##       --- Renderer      --- ##
+#################################
+
+class Engine3DGRUT:
```

Comment:
We should make it super clear that this is not mean for differentiable rendering / training of scenes. Folks will inevitably want that, and we don't want to disappoint :)